In [1]:
import pandas as pd


data = pd.read_csv("LoanStats_2018Q3.csv", skiprows=[0])
data.drop('member_id', axis=1, inplace=True)
data.drop('id', axis=1, inplace=True)

c:\users\guo\appdata\local\programs\python\python37\lib\site-packages\IPython\core\interactiveshell.py:3049: DtypeWarning: Columns (0,123,124,125,128,129,130,133,139,140,141) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [2]:
data

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  \
0         10000.0      10000.0          10000.0   36 months    6.67%   
1         35000.0      35000.0          35000.0   60 months   11.55%   
2          9600.0       9600.0           9600.0   36 months    7.84%   
3         18000.0      18000.0          18000.0   36 months    7.21%   
4         12000.0      12000.0          12000.0   36 months    8.46%   
5          5000.0       5000.0           5000.0   36 months    7.21%   
6         12550.0      12550.0          12550.0   36 months    8.46%   
7         20500.0      20500.0          20500.0   36 months   18.94%   
8         20000.0      20000.0          20000.0   36 months    6.67%   
9          5000.0       5000.0           5000.0   36 months   16.14%   
10        24000.0      24000.0          24000.0   60 months   15.02%   
11        15000.0      15000.0          15000.0   60 months   18.94%   
12         8500.0       8500.0           8500.0   36 months   14.47%   
13        35000.0      35000.0          35000.0   36 months    6.67%   
14        28000.0      28000.0          28000.0   60 months   14.47%   
15        20000.0      20000.0          20000.0   60 months   22.35%   
16        35000.0      35000.0          35000.0   60 months    7.84%   
17         6000.0       6000.0           6000.0   36 months    8.46%   
18         5025.0       5025.0           5025.0   36 months    7.84%   
19        10000.0      10000.0          10000.0   36 months   15.02%   
20        11200.0      11200.0          11200.0   36 months    7.84%   
21        20000.0      20000.0          20000.0   36 months    6.67%   
22        15000.0      15000.0          15000.0   60 months   15.02%   
23        20000.0      20000.0          20000.0   60 months   13.56%   
24        12000.0      12000.0          12000.0   36 months    6.11%   
25        12000.0      12000.0          12000.0   60 months   15.02%   
26         6400.0       6400.0           6400.0   36 months   10.08%   
27         9600.0       9600.0           9600.0   36 months   11.55%   
28        26000.0      26000.0          26000.0   36 months    6.11%   
29        24200.0      24200.0          24200.0   60 months   14.47%   
...           ...          ...              ...         ...      ...   
128166     9800.0       9800.0           9800.0   36 months   16.46%   
128167    20000.0      20000.0          20000.0   36 months   11.05%   
128168    14000.0      14000.0          14000.0   36 months   21.85%   
128169    40000.0      40000.0          39975.0   36 months    5.31%   
128170    20000.0      20000.0          19975.0   36 months   15.49%   
128171    30000.0      30000.0          29975.0   36 months    7.46%   
128172    36000.0      36000.0          35975.0   36 months    5.31%   
128173    28000.0      28000.0          27975.0   36 months    6.83%   
128174    21400.0      21400.0          21400.0   36 months   10.56%   
128175    24000.0      24000.0          23975.0   36 months   16.46%   
128176    38000.0      38000.0          38000.0   60 months    7.46%   
128177    35000.0      35000.0          35000.0   60 months    6.83%   
128178    24000.0      24000.0          24000.0   60 months   11.05%   
128179    10000.0      10000.0          10000.0   36 months   10.56%   
128180    17375.0      17375.0          17375.0   60 months   24.84%   
128181    20000.0      20000.0          20000.0   36 months   11.05%   
128182    10000.0      10000.0          10000.0   36 months   12.13%   
128183    10000.0      10000.0          10000.0   36 months   10.07%   
128184    40000.0      40000.0          40000.0   60 months   13.06%   
128185    15000.0      15000.0          15000.0   60 months   18.45%   
128186    15000.0      15000.0          15000.0   36 months   15.49%   
128187     5000.0       5000.0           5000.0   36 months   10.56%   
128188     5000.0       5000.0           5000.0   36 months   16.46%   
128189    12000.0      12000.0          12000.0 

In [3]:
data[-2:]

loan_amnt  funded_amnt  funded_amnt_inv term int_rate  installment  \
128194        NaN          NaN              NaN  NaN      NaN          NaN   
128195        NaN          NaN              NaN  NaN      NaN          NaN   

       grade sub_grade emp_title emp_length  ...  \
128194   NaN       NaN       NaN        NaN  ...   
128195   NaN       NaN       NaN        NaN  ...   

       hardship_payoff_balance_amount  hardship_last_payment_amount  \
128194                            NaN                           NaN   
128195                            NaN                           NaN   

       disbursement_method debt_settlement_flag debt_settlement_flag_date  \
128194                 NaN                  NaN                       NaN   
128195                 NaN                  NaN                       NaN   

       settlement_status  settlement_date  settlement_amount  \
128194               NaN              NaN                NaN   
128195               NaN              NaN                NaN   

       settlement_percentage settlement_term  
128194                   NaN             NaN  
128195                   NaN             NaN  

[2 rows x 143 columns]

In [4]:
print(data[-2:].isna().values.all())
print(data[-2:].isna().sum().sum())
print(data[-3:-2].isna().any().sum())

True
286
43


The last two rows are all missing values (only 43 are missing in the last third row), drop them. The last two rows are missing in the original excel dataset

In [5]:
data.drop([128194,128195], axis=0, inplace=True)

In [6]:
data

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  \
0         10000.0      10000.0          10000.0   36 months    6.67%   
1         35000.0      35000.0          35000.0   60 months   11.55%   
2          9600.0       9600.0           9600.0   36 months    7.84%   
3         18000.0      18000.0          18000.0   36 months    7.21%   
4         12000.0      12000.0          12000.0   36 months    8.46%   
5          5000.0       5000.0           5000.0   36 months    7.21%   
6         12550.0      12550.0          12550.0   36 months    8.46%   
7         20500.0      20500.0          20500.0   36 months   18.94%   
8         20000.0      20000.0          20000.0   36 months    6.67%   
9          5000.0       5000.0           5000.0   36 months   16.14%   
10        24000.0      24000.0          24000.0   60 months   15.02%   
11        15000.0      15000.0          15000.0   60 months   18.94%   
12         8500.0       8500.0           8500.0   36 months   14.47%   
13        35000.0      35000.0          35000.0   36 months    6.67%   
14        28000.0      28000.0          28000.0   60 months   14.47%   
15        20000.0      20000.0          20000.0   60 months   22.35%   
16        35000.0      35000.0          35000.0   60 months    7.84%   
17         6000.0       6000.0           6000.0   36 months    8.46%   
18         5025.0       5025.0           5025.0   36 months    7.84%   
19        10000.0      10000.0          10000.0   36 months   15.02%   
20        11200.0      11200.0          11200.0   36 months    7.84%   
21        20000.0      20000.0          20000.0   36 months    6.67%   
22        15000.0      15000.0          15000.0   60 months   15.02%   
23        20000.0      20000.0          20000.0   60 months   13.56%   
24        12000.0      12000.0          12000.0   36 months    6.11%   
25        12000.0      12000.0          12000.0   60 months   15.02%   
26         6400.0       6400.0           6400.0   36 months   10.08%   
27         9600.0       9600.0           9600.0   36 months   11.55%   
28        26000.0      26000.0          26000.0   36 months    6.11%   
29        24200.0      24200.0          24200.0   60 months   14.47%   
...           ...          ...              ...         ...      ...   
128164    40000.0      40000.0          39975.0   36 months    6.83%   
128165    14000.0      14000.0          13975.0   36 months   13.06%   
128166     9800.0       9800.0           9800.0   36 months   16.46%   
128167    20000.0      20000.0          20000.0   36 months   11.05%   
128168    14000.0      14000.0          14000.0   36 months   21.85%   
128169    40000.0      40000.0          39975.0   36 months    5.31%   
128170    20000.0      20000.0          19975.0   36 months   15.49%   
128171    30000.0      30000.0          29975.0   36 months    7.46%   
128172    36000.0      36000.0          35975.0   36 months    5.31%   
128173    28000.0      28000.0          27975.0   36 months    6.83%   
128174    21400.0      21400.0          21400.0   36 months   10.56%   
128175    24000.0      24000.0          23975.0   36 months   16.46%   
128176    38000.0      38000.0          38000.0   60 months    7.46%   
128177    35000.0      35000.0          35000.0   60 months    6.83%   
128178    24000.0      24000.0          24000.0   60 months   11.05%   
128179    10000.0      10000.0          10000.0   36 months   10.56%   
128180    17375.0      17375.0          17375.0   60 months   24.84%   
128181    20000.0      20000.0          20000.0   36 months   11.05%   
128182    10000.0      10000.0          10000.0   36 months   12.13%   
128183    10000.0      10000.0          10000.0   36 months   10.07%   
128184    40000.0      40000.0          40000.0   60 months   13.06%   
128185    15000.0      15000.0          15000.0   60 months   18.45%   
128186    15000.0      15000.0          15000.0   36 months   15.49%   
128187     5000.0       5000.0           5000.0 

In [7]:
headers = list(data)
print(headers)

['loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'term', 'int_rate', 'installment', 'grade', 'sub_grade', 'emp_title', 'emp_length', 'home_ownership', 'annual_inc', 'verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'url', 'desc', 'purpose', 'title', 'zip_code', 'addr_state', 'dti', 'delinq_2yrs', 'earliest_cr_line', 'inq_last_6mths', 'mths_since_last_delinq', 'mths_since_last_record', 'open_acc', 'pub_rec', 'revol_bal', 'revol_util', 'total_acc', 'initial_list_status', 'out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee', 'recoveries', 'collection_recovery_fee', 'last_pymnt_d', 'last_pymnt_amnt', 'next_pymnt_d', 'last_credit_pull_d', 'collections_12_mths_ex_med', 'mths_since_last_major_derog', 'policy_code', 'application_type', 'annual_inc_joint', 'dti_joint', 'verification_status_joint', 'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m', 'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since

In [8]:
# find out features with only one unique level and drop them

print(data.columns[data.nunique()==1])

Index(['pymnt_plan', 'policy_code', 'acc_now_delinq', 'num_tl_120dpd_2m',
       'num_tl_30dpd', 'tax_liens', 'hardship_flag', 'hardship_type',
       'hardship_reason', 'hardship_status', 'deferral_term',
       'hardship_length'],
      dtype='object')


In [9]:
print('the unique values in this feature are:', data['pymnt_plan'].unique())
print('the number of missing values is:', data['pymnt_plan'].isna().sum())

the unique values in this feature are: ['n']
the number of missing values is: 0


pymnt_plan has only one level, which is 'n', and has no missing values

In [10]:
headers_unique = list(data.columns[data.nunique()==1])

for header in headers_unique:
    print(header)
    print(data[header].unique())
    print(data[header].isna().sum())
    print()

pymnt_plan
['n']
0

policy_code
[1.]
0

acc_now_delinq
[0.]
0

num_tl_120dpd_2m
[ 0. nan]
3254

num_tl_30dpd
[0.]
0

tax_liens
[0.]
0

hardship_flag
['N']
0

hardship_type
[nan 'INTEREST ONLY-3 MONTHS DEFERRAL']
128181

hardship_reason
[nan 'NATURAL_DISASTER']
128181

hardship_status
[nan 'COMPLETED']
128181

deferral_term
[nan  3.]
128181

hardship_length
[nan  3.]
128181



128181 people have no hardship records becasue there are 128181 missing values in hardship-related features. It seems that 13 people experienced hardship and have completed it, but it is still an intuitively valuble predictor for indicating a person would likely default, so keep them and drop other unique level features first

In [11]:
print(headers_unique[:-6])
print(headers_unique[-6:])

['pymnt_plan', 'policy_code', 'acc_now_delinq', 'num_tl_120dpd_2m', 'num_tl_30dpd', 'tax_liens']
['hardship_flag', 'hardship_type', 'hardship_reason', 'hardship_status', 'deferral_term', 'hardship_length']


In [12]:
for header in headers_unique[:-6]:
    data.drop(header, axis=1, inplace=True)

Since all hardship features are indicating the same thing (3 months length of hardship term), only need to keep one hardship feature (the same for hardship_reason becuase it also only has one level). Keep hardship_status as a categorical feature and name missing values by 'No' as another category 

In [13]:
for header in headers_unique[-6:-3]:
    data.drop(header, axis=1, inplace=True)

for header in headers_unique[-2:]:
    data.drop(header, axis=1, inplace=True)

In [14]:
data['hardship_status'].fillna('No',inplace=True)

In [15]:
print(data['hardship_status'].unique())
print(data['hardship_status'].isna().sum())

['No' 'COMPLETED']
0


In [16]:
data.shape

(128194, 132)

In [17]:
# find out features that have missing values and impute them

data.isna().sum()

loan_amnt                                          0
funded_amnt                                        0
funded_amnt_inv                                    0
term                                               0
int_rate                                           0
installment                                        0
grade                                              0
sub_grade                                          0
emp_title                                      13437
emp_length                                     10387
home_ownership                                     0
annual_inc                                         0
verification_status                                0
issue_d                                            0
loan_status                                        0
url                                           128194
desc                                          128194
purpose                                            0
title                                         

In [18]:
headers_na = list(data.columns[data.isna().sum() > 0])
print('The number of features that have missing values are:',len(headers_na))
print('The features that have missing values are:')
print(headers_na)

The number of features that have missing values are: 54
The features that have missing values are:
['emp_title', 'emp_length', 'url', 'desc', 'dti', 'mths_since_last_delinq', 'mths_since_last_record', 'revol_util', 'last_pymnt_d', 'next_pymnt_d', 'last_credit_pull_d', 'mths_since_last_major_derog', 'annual_inc_joint', 'dti_joint', 'verification_status_joint', 'mths_since_rcnt_il', 'il_util', 'all_util', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util', 'mo_sin_old_il_acct', 'mths_since_recent_bc', 'mths_since_recent_bc_dlq', 'mths_since_recent_inq', 'mths_since_recent_revol_delinq', 'pct_tl_nvr_dlq', 'percent_bc_gt_75', 'revol_bal_joint', 'sec_app_earliest_cr_line', 'sec_app_inq_last_6mths', 'sec_app_mort_acc', 'sec_app_open_acc', 'sec_app_revol_util', 'sec_app_open_act_il', 'sec_app_num_rev_accts', 'sec_app_chargeoff_within_12_mths', 'sec_app_collections_12_mths_ex_med', 'sec_app_mths_since_last_major_derog', 'hardship_amount', 'hardship_start_date', 'hardship_end_date', 'payment_plan_start

impute hardship-related features for first

In [19]:
headers_na_hard = ['hardship_amount', 'hardship_start_date', 'hardship_end_date', 'payment_plan_start_date', 'hardship_dpd', 
                   'hardship_loan_status', 'orig_projected_additional_accrued_interest', 'hardship_payoff_balance_amount',
                   'hardship_last_payment_amount' ]

In [20]:
for header in headers_na_hard:
    print(header)
    print(data[header].unique())
    print(data[header].isna().sum())
    print()

hardship_amount
[   nan 203.67 378.9  126.27 119.59 243.56 159.94 241.9  134.81 194.82
  55.04 215.7  248.46 109.26]
128181

hardship_start_date
[nan 'Oct-2018' 'Nov-2018' 'Sep-2018']
128181

hardship_end_date
[nan 'Dec-2018' 'Jan-2019' 'Feb-2019']
128181

payment_plan_start_date
[nan 'Oct-2018' 'Nov-2018' 'Dec-2018']
128181

hardship_dpd
[nan  0. 18.]
128181

hardship_loan_status
[nan 'Issued' 'Current' 'Late (31-120 days)']
128181

orig_projected_additional_accrued_interest
[    nan  611.01 1136.7   378.81  358.77  730.68  479.82  725.7   404.43
  584.46  165.12  647.1   745.38  327.78]
128181

hardship_payoff_balance_amount
[     nan 40149.35 23090.22  8096.8   9649.7  12701.26 19071.77 19495.24
  9674.48 12492.2   5767.68 15414.46 23553.11  5932.38]
128181

hardship_last_payment_amount
[       nan 1.0100e+00 1.2100e+00 2.9283e+02 3.4643e+02 1.0000e-01
 1.9000e-01 4.7013e+02 2.0000e-01 3.2613e+02 1.9795e+02 2.4000e-01
 5.4237e+02 3.6648e+02]
128181



In [21]:
print(data['hardship_status'].unique())
print(data['hardship_status'].isna().sum())

['No' 'COMPLETED']
0


drop hardship related date features  

In [22]:
data.drop(labels=['hardship_start_date','hardship_end_date','payment_plan_start_date',
                              'hardship_dpd'],axis=1,inplace=True)

impute missing values in hardship_amount, orig_projected_additional_accrued_interest, hardship_payoff_balance_amount, and hardship_last_payment_amount by 0 (float), and name missing values in hardship_loan_status by 'No' as a new category

In [23]:
data['hardship_loan_status'].fillna('No',inplace=True)
data['hardship_amount'].fillna(0.,inplace=True)
data['orig_projected_additional_accrued_interest'].fillna(0.,inplace=True)
data['hardship_payoff_balance_amount'].fillna(0.,inplace=True)
data['hardship_last_payment_amount'].fillna(0.,inplace=True)

In [24]:
data.groupby(by=['hardship_loan_status']).size()

hardship_loan_status
Current                    9
Issued                     3
Late (31-120 days)         1
No                    128181
dtype: int64

In [25]:
data.shape

(128194, 128)

then, impute settlement-related features

In [26]:
print(data['debt_settlement_flag'].unique())
print(data['debt_settlement_flag'].isna().sum())

['N' 'Y']
0


In [27]:
headers_na_settle = ['debt_settlement_flag_date', 'settlement_status', 'settlement_date', 'settlement_amount', 
                     'settlement_percentage', 'settlement_term']

In [28]:
for header in headers_na_settle:
    print(header)
    print(data[header].unique())
    print(data[header].isna().sum())
    print()

debt_settlement_flag_date
[nan 'Feb-2019' 'Dec-2018' 'Jan-2019' 'Nov-2018']
128179

settlement_status
[nan 'ACTIVE' 'COMPLETE' 'BROKEN']
128179

settlement_date
[nan 'Feb-2019' 'Dec-2018' 'Jan-2019' 'Nov-2018']
128179

settlement_amount
[   nan 23506.  3422. 13175.  5180.  1059. 14333.  5694.  3332.  3852.
  3064. 12539. 16278.  5520. 21824.  5126.]
128179

settlement_percentage
[  nan 65.   65.01 60.   64.97 60.01 64.99]
128179

settlement_term
[nan 18.  6. 14.]
128179



In [29]:
data.groupby(by=['settlement_date']).size()

settlement_date
Dec-2018    3
Feb-2019    8
Jan-2019    2
Nov-2018    2
dtype: int64

In [30]:
data.groupby(by=['debt_settlement_flag_date']).size()

debt_settlement_flag_date
Dec-2018    2
Feb-2019    8
Jan-2019    3
Nov-2018    2
dtype: int64

In [31]:
data.groupby(by=['settlement_percentage']).size()

settlement_percentage
60.00    1
60.01    1
64.97    1
64.99    1
65.00    9
65.01    2
dtype: int64

In [32]:
data.groupby(by=['settlement_status']).size()

settlement_status
ACTIVE      13
BROKEN       1
COMPLETE     1
dtype: int64

In [33]:
settle_index = list(data.index[data['debt_settlement_flag_date'].isna()==False])
print(settle_index)

[15502, 27323, 54006, 65682, 72383, 80739, 82483, 88882, 90178, 106112, 112735, 120122, 123343, 124676, 125137]


In [34]:
charged_off_index = list(data.index[data['loan_status'] == 'Charged Off'])

In [35]:
print(any(i in charged_off_index for i in settle_index))
print(all(i in charged_off_index for i in settle_index))

True
False


Not all people on settlement plan are marked as charged off in loan_status 

In [36]:
settle_charge=[]
settle_not_in_charge=[]
for i in settle_index:
    if i in charged_off_index:
        settle_charge.append(i)
    else:
        settle_not_in_charge.append(i)
print(settle_charge)
print(settle_not_in_charge)

[82483, 88882, 90178, 106112, 120122, 125137]
[15502, 27323, 54006, 65682, 72383, 80739, 112735, 123343, 124676]


In [37]:
head = ['settlement_status', 'loan_status']

In [38]:
for i in settle_not_in_charge:
    print(data[head][i:i+1])

      settlement_status         loan_status
15502            ACTIVE  Late (31-120 days)
      settlement_status         loan_status
27323            ACTIVE  Late (31-120 days)
      settlement_status         loan_status
54006            ACTIVE  Late (31-120 days)
      settlement_status         loan_status
65682            ACTIVE  Late (31-120 days)
      settlement_status loan_status
72383            ACTIVE     Current
      settlement_status         loan_status
80739            ACTIVE  Late (31-120 days)
       settlement_status         loan_status
112735            ACTIVE  Late (31-120 days)
       settlement_status         loan_status
123343            ACTIVE  Late (31-120 days)
       settlement_status         loan_status
124676            ACTIVE  Late (31-120 days)


people on settlement plan are also in other loan status besides being charged off

In [39]:
data.groupby(by=['debt_settlement_flag']).size()

debt_settlement_flag
N    128179
Y        15
dtype: int64

drop settlement date related features

In [40]:
data.drop(labels=['debt_settlement_flag_date','settlement_date'],axis=1,inplace=True)

In [41]:
print(data['settlement_amount'].unique().dtype)
print(data['settlement_percentage'].unique().dtype)
print(data['settlement_term'].unique().dtype)

float64
float64
float64


In [42]:
data['settlement_status'].fillna('No',inplace=True)
data['settlement_amount'].fillna(0.,inplace=True)
data['settlement_percentage'].fillna(0.,inplace=True)
data['settlement_term'].fillna(0.,inplace=True)

In [43]:
data.shape

(128194, 126)

In [44]:
# inspect remaining features that have missing values

headers_na_rest = list(data.columns[data.isna().sum() > 0])
print(len(headers_na_rest),'remaining features to be imputed')
data[headers_na_rest].isna().sum()

39 remaining features to be imputed


emp_title                               13437
emp_length                              10387
url                                    128194
desc                                   128194
dti                                       307
mths_since_last_delinq                  71389
mths_since_last_record                 112487
revol_util                                129
last_pymnt_d                              141
next_pymnt_d                             8063
last_credit_pull_d                          4
mths_since_last_major_derog             98364
annual_inc_joint                       110409
dti_joint                              110409
verification_status_joint              112465
mths_since_rcnt_il                       4611
il_util                                 20514
all_util                                   29
avg_cur_bal                                12
bc_open_to_buy                           1647
bc_util                                  1696
mo_sin_old_il_acct                

impute features that are related to secondary or joint applicant

In [45]:
headers_na_sec = ['sec_app_earliest_cr_line', 'sec_app_inq_last_6mths','sec_app_mort_acc','sec_app_open_acc','sec_app_revol_util',
                 'sec_app_open_act_il','sec_app_num_rev_accts','sec_app_chargeoff_within_12_mths','sec_app_collections_12_mths_ex_med',
                 'sec_app_mths_since_last_major_derog','revol_bal_joint','verification_status_joint','dti_joint','annual_inc_joint']

In [46]:
for header in headers_na_sec:
    print(header)
    print(data[header].unique())
    print()

sec_app_earliest_cr_line
[nan 'Mar-1973' 'Apr-2002' 'Nov-2004' 'May-2006' 'Dec-2000' 'May-2011'
 'Sep-2002' 'Oct-1996' 'Nov-2005' 'Jan-2001' 'Jul-2005' 'Nov-2006'
 'Aug-2003' 'Jul-2001' 'Oct-2008' 'Sep-2016' 'Sep-2004' 'Jun-2007'
 'Jan-2014' 'Mar-2002' 'Sep-1993' 'Oct-2006' 'Feb-2005' 'May-2003'
 'Jul-2002' 'Jul-1994' 'Mar-2009' 'Apr-1987' 'Apr-1996' 'Jan-2002'
 'Jan-1994' 'Mar-2011' 'May-1986' 'Aug-2000' 'Apr-1975' 'Apr-2006'
 'Dec-2003' 'Jan-2000' 'Apr-2000' 'Sep-2011' 'Jan-2005' 'Apr-1991'
 'Jun-2003' 'Aug-1999' 'Dec-1998' 'Sep-2000' 'Mar-2016' 'Sep-2010'
 'Nov-2003' 'Dec-2011' 'Oct-2010' 'Jun-1999' 'Apr-1994' 'Jan-1976'
 'Oct-1995' 'Jun-2008' 'Mar-2004' 'Jan-2006' 'Dec-2008' 'Nov-2010'
 'Sep-2003' 'Jul-2010' 'Sep-2009' 'May-1978' 'Jun-2009' 'Mar-2001'
 'Apr-1993' 'Sep-1994' 'Jun-2010' 'Apr-2014' 'Aug-1993' 'Dec-2010'
 'Oct-2004' 'Jul-2007' 'Oct-2009' 'Nov-2007' 'Dec-2007' 'Dec-1999'
 'Sep-2001' 'Feb-1986' 'Aug-1994' 'Jan-2011' 'Aug-1997' 'Sep-2007'
 'Dec-2006' 'Apr-2004' 'Oct-2013'

In [47]:
data[headers_na_sec].isna().sum()

sec_app_earliest_cr_line               110409
sec_app_inq_last_6mths                 110409
sec_app_mort_acc                       110409
sec_app_open_acc                       110409
sec_app_revol_util                     110719
sec_app_open_act_il                    110409
sec_app_num_rev_accts                  110409
sec_app_chargeoff_within_12_mths       110409
sec_app_collections_12_mths_ex_med     110409
sec_app_mths_since_last_major_derog    122582
revol_bal_joint                        110409
verification_status_joint              112465
dti_joint                              110409
annual_inc_joint                       110409
dtype: int64

sec_app_revol_util, sec_app_mths_since_last_major_derog, and verification_status_joint have different number of NA than others. sec_app_mths_since_last_major_derog is ok becuase it may not apply to some secondary applicants. examine the other two features

In [48]:
sec_index_base1 = list(data.index[data['sec_app_earliest_cr_line'].isna()==False])
sec_index_base2 = list(data.index[data['sec_app_mort_acc'].isna()==False])

In [49]:
all(i in sec_index_base1 for i in sec_index_base2)

True

confirm indexes of features that have 110409 NAs matching first, and use the Non-NAs' indexes as baseline for comparison; confirm if all indexes of Non-NAs in sec_app_revol_util and verification_status_joint are in baseline

In [50]:
sec_index_revol = list(data.index[data['sec_app_revol_util'].isna()==False])
sec_index_veri = list(data.index[data['verification_status_joint'].isna()==False])

In [51]:
sec_diff_revol=[]
sec_diff_veri=[]
for i in sec_index_base1:
    if i not in sec_index_revol:
        sec_diff_revol.append(i)

for i in sec_index_base1:
    if i not in sec_index_veri:
        sec_diff_veri.append(i)
        
print(len(sec_index_base1)-len(sec_index_revol)==len(sec_diff_revol))     
print(len(sec_index_base1)-len(sec_index_veri)==len(sec_diff_veri))  

True
True


confirmed all indexes in sec_app_revol_util and verification_status_joint are in baseline

In [52]:
data.groupby(by=['verification_status_joint']).size()

verification_status_joint
Not Verified       6827
Source Verified    5601
Verified           3301
dtype: int64

In [53]:
any(data['sec_app_revol_util']==0.)

True

In [54]:
print(data['sec_app_revol_util'].median())
print(data['sec_app_revol_util'].mean())
print(data['sec_app_revol_util'].mode())

58.4
56.37079256080114
0    0.0
dtype: float64


Since there are data=0 in sec_app_revol_util, cannot use 0 to impute missing values of people not having secondary applicant, use -1 instead. Use median to impute missing values of people having secondary applicant. 
Use 'Not Verified' to impute verification_status_joint for people having secondary applicant, and 'No' to impute people not having secondary applicant

In [55]:
for i in sec_diff_revol:
    data['sec_app_revol_util'][i:i+1].fillna(data['sec_app_revol_util'].median(),inplace=True)

In [56]:
data['sec_app_revol_util'].fillna(-1.,inplace=True)

In [57]:
for i in sec_diff_veri:
    data['verification_status_joint'][i:i+1].fillna('Not Verified',inplace=True)

In [58]:
data['verification_status_joint'].fillna('No',inplace=True)

In [59]:
data.groupby(by=['verification_status_joint']).size()

verification_status_joint
No                 110409
Not Verified         8883
Source Verified      5601
Verified             3301
dtype: int64

In [60]:
data['sec_app_mths_since_last_major_derog'].fillna(-1.,inplace=True)

now impute the rest secondary applicant related features by -1 because they are all numeric predictors, except for sec_app_earliest_cr_line because it is a date type predictor

In [61]:
for header in headers_na_sec:
    if header not in ['sec_app_revol_util','verification_status_joint','sec_app_mths_since_last_major_derog', 'sec_app_earliest_cr_line']:
        data[header].fillna(-1.,inplace=True)

In [62]:
data['sec_app_earliest_cr_line'].dtype

dtype('O')

now deal with date type predictor sec_app_earliest_cr_line; impute it with another category, say 'Jan-1900'

In [63]:
data['sec_app_earliest_cr_line'].fillna('Jan-1900',inplace=True)

In [64]:
# Again, inspect remaining features that have missing values

headers_na_rest1 = list(data.columns[data.isna().sum() > 0])
print(len(headers_na_rest1),'remaining features to be imputed')
data[headers_na_rest1].isna().sum()

25 remaining features to be imputed


emp_title                          13437
emp_length                         10387
url                               128194
desc                              128194
dti                                  307
mths_since_last_delinq             71389
mths_since_last_record            112487
revol_util                           129
last_pymnt_d                         141
next_pymnt_d                        8063
last_credit_pull_d                     4
mths_since_last_major_derog        98364
mths_since_rcnt_il                  4611
il_util                            20514
all_util                              29
avg_cur_bal                           12
bc_open_to_buy                      1647
bc_util                             1696
mo_sin_old_il_acct                  4611
mths_since_recent_bc                1550
mths_since_recent_bc_dlq          102769
mths_since_recent_inq              16016
mths_since_recent_revol_delinq     91113
pct_tl_nvr_dlq                         1
percent_bc_gt_75

drop url and desc becuase the two columns are entirely missing

In [65]:
data.drop(labels=['url','desc'], axis=1, inplace=True)

In [66]:
print(data['emp_title'].unique())
print(data['emp_title'].nunique())

['Customer Service Rep' 'Associate Director, IT' 'Branch manager ' ...
 'Medical Support Staff' 'Host/cashier' 'jewelry']
44627


drop emp_title because of too many levels

In [67]:
data.drop('emp_title', axis=1,inplace=True)

In [68]:
data.shape

(128194, 123)

inspect features that have less than 1000 missing values, impute numeric predictor by mean or median, categorical by mode

In [69]:
data['dti'].dtype

dtype('float64')

In [70]:
print(data['dti'].mean())
print(data['dti'].median())

19.81038401088461
17.81


In [71]:
data['dti'].fillna(data['dti'].mean(),inplace=True)

In [72]:
print(data['revol_util'].dtype)
print(data['revol_util'][:10])

object
0    24.1%
1    48.8%
2    10.8%
3    13.2%
4    26.7%
5       8%
6     3.2%
7     3.8%
8    35.9%
9    74.1%
Name: revol_util, dtype: object


In [73]:
data['revol_util'] = data['revol_util'].str.rstrip('%').astype('float')

In [74]:
print(data['revol_util'][:10])

0    24.1
1    48.8
2    10.8
3    13.2
4    26.7
5     8.0
6     3.2
7     3.8
8    35.9
9    74.1
Name: revol_util, dtype: float64


In [75]:
print(data['revol_util'].mean())
print(data['revol_util'].median())

44.45021668683872
42.7


In [76]:
data['revol_util'].fillna(data['revol_util'].mean(),inplace=True)

In [77]:
print(data['last_pymnt_d'].dtype)

object


In [78]:
data.groupby(by=['last_pymnt_d']).size()

last_pymnt_d
Aug-2018       837
Dec-2018      1678
Feb-2019    116381
Jan-2019      4633
Jul-2018       222
Nov-2018      1613
Oct-2018      1562
Sep-2018      1127
dtype: int64

In [79]:
data['last_pymnt_d'].fillna('Feb-2019',inplace=True)

In [80]:
print(data['last_credit_pull_d'].dtype)

object


In [81]:
data.groupby(by=['last_credit_pull_d']).size()

last_credit_pull_d
Aug-2018       565
Dec-2018      1048
Feb-2019    122163
Jan-2019      1803
Jul-2018       335
Jun-2018        53
Nov-2018       876
Oct-2018       687
Sep-2018       660
dtype: int64

In [82]:
data['last_credit_pull_d'].fillna('Feb-2019',inplace=True)

In [83]:
print(data['all_util'].dtype)

float64


In [84]:
print(data['all_util'].mean())
print(data['all_util'].median())

54.353747122849455
55.0


In [85]:
data['all_util'].fillna(data['all_util'].mean(),inplace=True)

In [86]:
print(data['avg_cur_bal'].dtype)

float64


In [87]:
print(data['avg_cur_bal'].mean())
print(data['avg_cur_bal'].median())

13695.71962522039
7080.0


In [88]:
data['avg_cur_bal'].fillna(data['avg_cur_bal'].median(),inplace=True)

In [89]:
print(data['pct_tl_nvr_dlq'].dtype)

float64


In [90]:
print(data['pct_tl_nvr_dlq'].mean())
print(data['pct_tl_nvr_dlq'].median())

94.58931610930394
100.0


In [91]:
data['pct_tl_nvr_dlq'].fillna(data['pct_tl_nvr_dlq'].mean(),inplace=True)

inspect features that have more than 1000 missing values; use other features to predict missing values

In [92]:
headers_na_rest2 = list(data.columns[data.isna().sum() > 0])
print(len(headers_na_rest2),'remaining features to be imputed')
data[headers_na_rest2].isna().sum()

15 remaining features to be imputed


emp_length                         10387
mths_since_last_delinq             71389
mths_since_last_record            112487
next_pymnt_d                        8063
mths_since_last_major_derog        98364
mths_since_rcnt_il                  4611
il_util                            20514
bc_open_to_buy                      1647
bc_util                             1696
mo_sin_old_il_acct                  4611
mths_since_recent_bc                1550
mths_since_recent_bc_dlq          102769
mths_since_recent_inq              16016
mths_since_recent_revol_delinq     91113
percent_bc_gt_75                    1649
dtype: int64

In [93]:
for header in headers_na_rest2:
    if data[header].dtype != "float64":
        print(header)
        print(data[header].dtype)

emp_length
object
next_pymnt_d
object


In [94]:
data.shape

(128194, 123)

In [95]:
data.to_csv("15_more_to_be_imputed_data.csv",index=False)